### ISSUE TITLE
Cygwin file permission handling is perplexing if not outright broken

### ISSUE THREAD

**Issue by krader1961** [att/ast#1281]
The `path.sh` unit test fails several individual tests on Cygwin. For example,

```
<E> path[290]: unreadable empty script should fail
<E> expect: |126|
<E> actual: |0|
```

AFAICT all the failures are because Cygwin file permission handling is broken. Comparing the current ksh code with the ksh93u code it is unclear why the latter handles Cygwin correctly while the current code does not. The basic problem is that `open()` succeeds even if we don't have permission to open the file (e.g., open readonly and no read access bits are set). And `execve()` returns failure with errno == ENOEXEC rather than the expected EACCES when the file doesn't have execute permission.

In all likelihood the extremely confusing "intercept" macros we removed that made calls to APIs like `open()` actually call a wrapper function (e.g., `ast_open()`) account for the difference in behavior.

**Comment by krader1961** [comment]
Cygwin file permission handling on local filesystems essentially ignores usual UNIX semantics: http://cygwin.com/cygwin-ug-net/using-filemodes.html

**Comment by krader1961** [comment]
And ksh93u+ handles some situations correctly but not others. For example, trying to execute an empty file with only execute (no read) permission should also result in an exit status of 126. But ksh93u+ does open the file and return status zero. It's not at all clear we should complicate the code by special-casing `sh_open()` on Cygwin. Especially since the `access()` function returns the wrong result in this situation. I'm going to sleep on this but I am tempted to just skip the failing tests on Cygwin rather than complicate the code to try and provide UNIX behavior for these fringe cases. Especially since the Cygwin emulation layer doesn't do so. It really isn't our responsibility to do so.

**Comment by krader1961** [comment]
FWIW, I ran the *path.sh* unit test using the ksh93u+ binary that ships with macOS. It fails thusly:

```
<E> path[152]: exit status of non-executable is wrong
<E> expect: |126|
<E> actual: |127|
```

I can't actually test the behavior of the real ksh on Cygwin because it only ships `mksh` which `ksh` symlinks to. And mksh fails the *path.sh* unit test in novel ways.

**Comment by krader1961** [comment]
Another FWIW: I substituted the ksh93v- binary for the current ksh binary when running unit tests. There are 39 unit test failures on my macOS system. Many unit tests had multiple individual test failures. Many of those test failures are because of deliberate changes such as eliminating the `grep` builtin where we've adapted the tests to those changes. But many of the failures are due to actual bugs in the ksh93v- code that we have fixed.
